ethcore/src/block_queue.rs

```diff
@@ -285,18 +285,24 @@ impl BlockQueue {
 	}
 
 	/// Mark given block and all its children as bad. Stops verification.
-	pub fn mark_as_bad(&mut self, hash: &H256) {
+	pub fn mark_as_bad(&mut self, hashes: &[H256]) {
 		let mut verification_lock = self.verification.lock().unwrap();
+		let mut processing = self.processing.write().unwrap();
+
 		let mut verification = verification_lock.deref_mut();
-		verification.bad.insert(hash.clone());
-		self.processing.write().unwrap().remove(&hash);
+
+		verification.bad.reserve(hashes.len());
+		for hash in hashes {
```

Comment:
We would need to clone everything in `L352` anyway, because we are using it to mark blocks as bad/good but then it's moved with `SyncMessage`. I think there was a todo to try (Cow)[https://doc.rust-lang.org/std/borrow/enum.Cow.html] in `SyncMessage` instead - I can try that.